documentation/specs/task-isolation-and-dependencies.md

```diff
@@ -0,0 +1,38 @@
+# Task isolation
+## Problem definition
+Tasks in MSBuild are dynamically loaded assemblies with potentially separate and colliding dependency trees. Currently MSBuild on .NET Core has no isolation between tasks and as such only one version of any given assembly can be loaded. Prime example of this is Newtonsoft.Json which has multiple versions, but all the tasks must agree on it to work.
+This problem is also described in microsoft/msbuild#1754.
+
+## Possible solution
+Use [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.2) (ALC) to provide binding isolation for tasks. Each task would be loaded into its own ALC instance.
+* The ALC would resolve all dependencies of the tasks (see dependency resolution below)
+* ALC would fallback to the Default for dependencies which the task doesn't carry with itself (Framework and so on)
+* ALC would probably have to forcefully fallback for MSBuild assemblies since by default lot of tasks will carry these, but the system requires for the MSBuild assemblies to be shared.
```

Comment:
What does "carry" mean in this context? I wouldn't expect tasks to _redistribute_ MSBuild assemblies, but they will depend on them. How does that impact their `.deps.json`.

Reply:
It means the task has a second copy of some of the MSBuild assemblies with it. I agree that well behaved tasks should not do this. The problem is tooling defaults. If one creates a classlib and builds it, all the dependencies are copied into the output. There are ways to mark the dependencies as "don't copy to output", but it's something which has to be done explicitly. If somebody forgets, the assembly will be in the output. We've ran into this with "plugin load" samples and it's relatively hard to diagnose when that happens. It's better to try to avoid this in the host if possible.